sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -652,6 +653,299 @@ case class MapObjects private(
+        case MapType(_, _, _) => makeCopyIfInstanceOf(classOf[UnsafeMapData], genFunction.value)
+        case _ => genFunction.value
+      }
+    val genKeyFunctionValue = genFunctionValue(keyLambdaFunction, genKeyFunction)
+    val genValueFunctionValue = genFunctionValue(valueLambdaFunction, genValueFunction)
+
+    def loopNullCheck(genInputData: ExprCode, inputDataType: DataType,
+                      loopIsNull: String, loopValue: String) =
+      inputDataType match {
+        case _: ArrayType => s"$loopIsNull = ${genInputData.value}.isNullAt($loopIndex);"
+        // The element of primitive array will never be null.
+        case ObjectType(cls) if cls.isArray && cls.getComponentType.isPrimitive =>
+          s"$loopIsNull = false"
+        case _ => s"$loopIsNull = $loopValue == null;"
+      }
+    val keyLoopNullCheck =
+      loopNullCheck(genKeyInputData, keyInputDataType, keyLoopIsNull, keyLoopValue)
+    val valueLoopNullCheck =
+      loopNullCheck(genValueInputData, valueInputDataType, valueLoopIsNull, valueLoopValue)
+
+    val constructBuilder = collClass match {
+      // Scala Map
+      case cls if classOf[scala.collection.Map[_, _]].isAssignableFrom(cls) =>
+        val builderClass = classOf[Builder[_, _]].getName
+        s"""
+          $builderClass $builderValue = ${collClass.getName}$$.MODULE$$.newBuilder();
+          $builderValue.sizeHint($dataLength);
+        """
+      // Java Map, AbstractMap => HashMap
+      case cls if classOf[java.util.Map[_, _]] == cls ||
+        classOf[java.util.AbstractMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.HashMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass($dataLength);"
+      // Java SortedMap, NavigableMap => TreeMap
+      case cls if classOf[java.util.SortedMap[_, _]] == cls ||
+        classOf[java.util.NavigableMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.TreeMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java ConcurrentMap => ConcurrentHashMap
+      case cls if classOf[java.util.concurrent.ConcurrentMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.concurrent.ConcurrentHashMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java ConcurrentNavigableMap => ConcurrentSkipListMap
+      case cls if classOf[java.util.concurrent.ConcurrentNavigableMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.concurrent.ConcurrentSkipListMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java concrete Map implementation
+      case cls =>
+        val builderClass = classOf[java.util.Map[_, _]].getName
+        // Check for constructor with capacity specification
+        if (Try(cls.getConstructor(Integer.TYPE)).isSuccess) {
+          s"$builderClass $builderValue = new ${cls.getName}($dataLength);"
```

Comment:
I think we can use the customer map type as the type declaration, then we don't need the cast in https://github.com/apache/spark/pull/16986/files#diff-e436c96ea839dfe446837ab2a3531f93R901